lib/check-meta.nix

```diff
@@ -0,0 +1,200 @@
+
+        b) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
+          { allow${allow_attr} = true; }
+        to ~/.config/nixpkgs/config.nix.
+      '';
+
+    remediate_insecure = attrs:
+      ''
+
+        Known issues:
+
+      '' + (lib.fold (issue: default: "${default} - ${issue}\n") "" attrs.meta.knownVulnerabilities) + ''
+
+          You can install it anyway by whitelisting this package, using the
+          following methods:
+
+          a) for `nixos-rebuild` you can add ‘${attrs.name or "«name-missing»"}’ to
+             `nixpkgs.config.permittedInsecurePackages` in the configuration.nix,
+             like so:
+
+               {
+                 nixpkgs.config.permittedInsecurePackages = [
+                   "${attrs.name or "«name-missing»"}"
+                 ];
+               }
+
+          b) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
+          ‘${attrs.name or "«name-missing»"}’ to `permittedInsecurePackages` in
+          ~/.config/nixpkgs/config.nix, like so:
+
+               {
+                 permittedInsecurePackages = [
+                   "${attrs.name or "«name-missing»"}"
+                 ];
+               }
+
+        '';
+
+    throwEvalHelp = { reason , errormsg ? "" }:
+      throw (''
+        Package ‘${attrs.name or "«name-missing»"}’ in ${pos_str} ${errormsg}, refusing to evaluate.
+
+        '' + ((builtins.getAttr reason remediation) attrs));
+
+    metaTypes = with lib.types; rec {
+      # These keys are documented
+      description = str;
+      longDescription = str;
+      branch = str;
+      homepage = str;
+      downloadPage = str;
+      license = either (listOf lib.types.attrs) (either lib.types.attrs str);
+      maintainers = listOf str;
+      priority = int;
+      platforms = listOf str;
+      hydraPlatforms = listOf str;
+      broken = bool;
+
+      # Weirder stuff that doesn't appear in the documentation?
+      version = str;
+      tag = str;
+      updateWalker = bool;
+      executables = listOf str;
+      outputsToInstall = listOf str;
+      position = str;
+      repositories = attrsOf str;
+      isBuildPythonPackage = platforms;
+      schedulingPriority = str;
+      downloadURLRegexp = str;
+      isFcitxEngine = bool;
+      isIbusEngine = bool;
+    };
+
+    checkMetaAttr = k: v:
+      if metaTypes?${k} then
+        if metaTypes.${k}.check v then null else "key '${k}' has a value ${v} of an invalid type ${builtins.typeOf v}; expected ${metaTypes.${k}.description}"
+      else "key '${k}' is unrecognized; expected one of: \n\t      [${lib.concatMapStringsSep ", " (x: "'${x}'") (lib.attrNames metaTypes)}]";
+    checkMeta = meta: if shouldCheckMeta then lib.remove null (lib.mapAttrsToList checkMetaAttr meta) else [];
+
+    # Check if a derivation is valid, that is whether it passes checks for
+    # e.g brokenness or license.
+    #
+    # Return { valid: Bool } and additionally
```

Comment:
It's more of a:

```haskell
data  Result = Valid | Invalid { reason :: String, ErrorMsg :: String }
```

Where `valid: Bool` is the tag in the tagged union, and thus must always be present.